handler.go

```diff
@@ -436,12 +451,41 @@ func (h *Handler) handleFetch(ctx *Context, req *protocol.FetchRequest) *protoco
 				log.Printf("invalid partition request: %d", fetchTopic.Partitions[0].Partition)
 				return
 			}
-			fetchOffset := fetchTopic.Partitions[0].FetchOffset
+
+			// If an offset is being requested that we don't yet have available, a partition
+			// response with no records should be returned.
+			startingOffset := fetchTopic.Partitions[0].FetchOffset
+			if startingOffset > maxOffsetAvailable {
+				responseChan <- &protocol.FetchTopicResponse{
+					Topic: fetchTopic.Topic,
+					PartitionResponses: []*protocol.FetchPartitionResponse{
+						{
+							Partition:           0,
+							ErrorCode:           0,
+							HighWatermark:       math.MaxInt64,
+							LastStableOffset:    math.MaxInt64,
+							AbortedTransactions: nil,
+							RecordSet:           nil,
+						},
+					},
+				}
+				return
+			}
 
 			// Build RecordSet to respond to this topic.
 			var buf bytes.Buffer
 			for x := 0; x < h.config.MaxMessagesPerTopic; x++ {
-				offset, data, err := mp(deadlineCtx, fetchOffset)
+				// Each message in the RecordSet batch is from the "next" offset as we build the
+				// batch.
+				thisOffset := startingOffset + int64(x)
+
+				// If we've exceeded the offsets for which there are messages available, there are
+				// no more available messages.
+				if thisOffset > maxOffsetAvailable {
```

Comment:
I haven't actually worked through the math, but my intuition says there might be an off-by-one error here.

If `maxOffsetAvailable` is zero should any messages be returned? This feels like a situation where you want to only return offsets _less than_ the maximum, and I think as currently written you'll return offsets _less than or equal to_ the maximum?

Reply:
Yeah this wasn't quite right. Thinking about it more I have come to the conclusion that it is a little easier to consider it in terms of the the number of _records_ available rather than the maximum _offset_ available. Offsets are zero-indexed, so a request for offset 0 will return a record if there is a single record available. Trying to represent the maximum offset available as I was doing would make it strange to simulate there being no records available...I guess `maxOffsetAvailable` would need to be `-1` for this, but I think that would break the offsets request handler.

So I refactored the function to be `RecordsAvailable` and added some additional comments explaining these nuances.